### PR TITLE
Finished lecture 1 exercises

### PR DESCRIPTION
**Pull request by axiomsbane**
### Solutions for Lecture 1

cc @chshersh

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -54,7 +55,8 @@ Explanation: @sumOfSquares 3 4@ should be equal to @9 + 16@ and this
 is 25.
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-sumOfSquares x y = error "TODO!"
+sumOfSquares :: Num a => a -> a -> a
```
Comment (chshersh) [inline src/Lecture1.hs:58]: `Num` is explained in the following lectures, so you are one step ahead 😸
In here `Int -> Int -> Int` or `Integer -> Integer -> Integer` would also work, as a more specific type 🙂

File: src/Lecture1.hs [hunk 2]
```diff
@@ -67,7 +69,10 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Integral a => a -> a
+lastDigit n
+    | n >= 0 && n<= 9 = n
+    | otherwise = lastDigit (mod (abs n) 10)
```
Comment (chshersh) [inline src/Lecture1.hs:75]: You don't need a recursive call to `lastDigit` and a bounds check. `mod (abs n) 10` already returns the expected result 👍🏻

File: src/Lecture1.hs [hunk 3]
```diff
@@ -108,7 +119,21 @@ and finds a sum of the numbers inside this string.
 
 The string contains only spaces and/or numbers.
 -}
-strSum str = error "TODO"
+strSum :: String -> Int
+strSum str = calcIt 0 str
+    where calcIt summ [] = summ
+          calcIt summ str1 = calcIt (summ + getInt (takeWhile charu (dropWhile space str1))) (dropWhile charu (dropWhile space str1))
```
Comment (chshersh) [inline src/Lecture1.hs:125]: You can use standard functions to simplify the implementation of `strSum`:

* `words`: to split a string into a list of space-separated words
* `read`: to parse an `Int` from `String`
* `sum`: to find a sum of all numbers in the list

File: src/Lecture1.hs [hunk 4]
```diff
@@ -123,4 +148,7 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Int -> [Int] -> [Char]
+lowerAndGreater n list = show n ++ " is greater than " ++ gt ++ " elements and lower than " ++ lt ++ " elements"
+    where lt = show $ length $ filter (> n) list
+          gt = show $ length $ filter (< n) list
```
Comment (chshersh) [inline src/Lecture1.hs:154]: That's correct! 💯 

However, with this solution, you will walk through the list twice.
You can write it more performant with only one traversal of the list 🙂
It's desirable to write functions that traverse lists only a single time and consume elements only once. In that case, the composition of such functions will have nice streaming and lazy properties.

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 3 above)
Comment (axiomsbane) [reply src/Lecture1.hs:125]: Yupp... I have to start forcing myself to use more prelude functions 😅.

**Pull request merged**